File: scripts/rrt.py
"""
ESE 680
RRT assignment
Author: Hongrui Zheng

This file contains the class definition for tree nodes and RRT
Before you start, please read: https://arxiv.org/pdf/1105.1186.pdf
"""
import numpy as np
from numpy import linalg as LA
import math
from copy import deepcopy

import rospy
from sensor_msgs.msg import LaserScan
from geometry_msgs.msg import PoseStamped
from geometry_msgs.msg import PointStamped
from geometry_msgs.msg import Pose
from geometry_msgs.msg import Point
from ackermann_msgs.msg import AckermannDriveStamped
from nav_msgs.msg import OccupancyGrid
from nav_msgs.msg import Path
from std_msgs.msg import Float64MultiArray
from tf import transform_listener

# class def for tree nodes
class Node(object):
    def __init__(self):
        self.x = None
        self.y = None
        self.parent = None
        self.cost = None # only used in RRT*
        self.is_root = false

# class def for RRT
class RRT(object):
    def __init__(self):
        # get params from yaml file
        # topics, not saved as attributes
        sim_drive_topic = rospy.get_param('sim_drive_topic')
        drive_topic = rospy.get_param('drive_topic')
        pf_topic = rospy.get_param('pf_pose_topic')
        static_topic = rospy.get_param('static_layer_topic')
        dynamic_topic = rospy.get_param('dynamic_layer_topic')
        env_topic = rospy.get_param('env_layer_topic')
        # wpt_topic = rospy.get_param('wayoint_topic')
        path_topic = rospy.get_param('path_topic')
        tree_topic = rospy.get_param('tree_topic')

        use_sim = rospy.get_param('use_sim')
        # you could add your own parameters to the rrt_params.yaml file,
        # and get them here as class attributes as shown above.

        # subscribers
        rospy.Subscriber(pf_topic, PoseStamped, self.pf_callback)
        rospy.Subscriber(static_topic, OccupancyGrid, self.static_callback)
        rospy.Subscriber(dynamic_topic, OccupancyGrid, self.dynamic_callback)
        rospy.Subscriber(env_topic, OccupancyGrid, self.env_callback)

        # publishers
        # drive
        if use_sim:
            self.drive_pub = rospy.Publisher(drive_topic, AckermannDriveStamped, queue_size=10)
        else:
            self.drive_pub = rospy.Publisher(sim_drive_topic, AckermannDriveStamped, queue_size=10)
        # visualization
        self.tree_pub = rospy.Publisher(tree_topic, Float64MultiArray, queue_size=10)
        # self.wpt_pub = rospy.Publisher(wpt_topic, Point, queue_size=10)
        self.path_pub = rospy.Publisher(path_topic, Path, queue_size=10)

        # class attributes
        # TODO: load list of waypoints from csv
        self.waypoints = None
        # current occupancy grid layers
        self.current_env_layer = None
        self.current_static_layer = None
        self.current_dynamic_layer = None
        # map metadata
        self.map_resolution = None
        self.map_width = None
        self.map_height = None
        self.origin_x = None
        self.origin_y = None

        # wait for all topics to start publishing before starting RRT
        # wait for map metadata
        map_metadata_msg = rospy.wait_for_message('/map_metadata', MapMetaData)
        self.map_resolution = map_metadata_msg.resolution
        self.map_width = map_metadata_msg.width
        self.map_height = map_metadata_msg.height
        self.origin_x = map_metadata_msg.origin.position.x
        self.origin_y = map_metadata_msg.origin.position.y
        rospy.log_info('Map Metadata Loaded')
        dynamic_msg = rospy.wait_for_message(dynamic_topic, OccupancyGrid)
        static_msg = rospy.wait_for_message(static_topic, OccupancyGrid)
        env_msg = rospy.wait_for_message(env_topic, OccupancyGrid)
        self.dynamic_callback(dynamic_msg)
        self.static_callback(static_msg)
        self.env_callback(env_msg)
        rospy.log_info('Occupancy Grid layers loaded.')

    def env_callback(self, env_msg):
        self.current_env_layer = deepcopy(env_msg)

    def static_callback(self, static_msg):
        self.current_static_layer = deepcopy(static_msg)

    def dynamic_callback(self, dynamic_msg):
        self.current_dynamic_layer = deepcopy(dynamic_msg)

    def pf_callback(self, pose_msg):
        """
        The pose callback when subscribed to particle filter's inferred pose
        Here is where the main RRT loop happens

        Args: 
            pose_msg (PoseStamped): incoming message from subscribed topic
        Returns:

        """
        tree = [] # tree as list of Nodes

        path_msg = Path()
        # fill in the path_msg with the path rrt finds
        self.path_pub.publish(path_msg)
        # visualization
        self.pub_tree(tree)
        return None

    def sample(self):
        """
        This method should randomly sample the free space, and returns a viable point

        Args:
        Returns:
            (x, y) (float float): a tuple representing the sampled point

        """
        x = None
        y = None
        return (x, y)

    def nearest(self, tree, sampled_point):
        """
        This method should return the nearest node on the tree to the sampled point

        Args:
            tree ([]): the current RRT tree
            sampled_point (tuple of (float, float)): point sampled in free space
        Returns:
            nearest_node (int): index of neareset node on the tree
        """
        nearest_node = 0
        return nearest_node

    def steer(self, nearest_node, sampled_point):
        """
        This method should return a point in the viable set such that it is closer 
        to the nearest_node than sampled_point is.

        Args:
            nearest_node (Node): nearest node on the tree to the sampled point
            sampled_point (tuple of (float, float)): sampled point
        Returns:
            new_node (Node): new node created from steering
        """
        new_node = None
        return new_node

    def check_collision(self, nearest_node, new_node):
        """
        This method should return whether the path between nearest and new_node is
        collision free.

        Args:
            nearest (Node): nearest node on the tree
            new_node (Node): new node from steering
        Returns:
            collision (bool): whether the path between the two nodes are in collision
                              with the occupancy grid
        """
        return True

    def is_goal(self, latest_added_node, goal_x, goal_y):
        """
        This method should return whether the latest added node is close enough
        to the goal.

        Args:
            latest_added_node (Node): latest added node on the tree
            goal_x (double): x coordinate of the current goal
            goal_y (double): y coordinate of the current goal
        Returns:
            close_enough (bool): true if node is close enoughg to the goal
        """
        return False

    def find_path(self, tree, latest_added_node):
        """
        This method returns a path as a list of Nodes connecting the starting point to
        the goal once the latest added node is close enough to the goal

        Args:
            tree ([]): current tree as a list of Nodes
            latest_added_node (Node): latest added node in the tree
        Returns:
            path ([]): valid path as a list of Nodes
        """
        path = []
        return path



    # The following methods are needed for RRT* and not RRT
    def cost(self, tree, node):
        """
        This method should return the cost of a node

        Args:
            node (Node): the current node the cost is calculated for
        Returns:
            cost (float): the cost value of the node
        """
        return 0

    def line_cost(self, n1, n2):
        """
        This method should return the cost of the straight line between n1 and n2

        Args:
            n1 (Node): node at one end of the straight line
            n2 (Node): node at the other end of the straint line
        Returns:
            cost (float): the cost value of the line
        """
        return 0

    def near(self, tree, node):
        """
        This method should return the neighborhood of nodes around the given node

        Args:
            tree ([]): current tree as a list of Nodes
            node (Node): current node we're finding neighbors for
        Returns:
            neighborhood ([]): neighborhood of nodes as a list of Nodes
        """
        neighborhood = []
        return neighborhood


    def pub_tree(self, tree):
        """
        This method publishes the tree as a Float64MultiArray message for visualization
        """
        tree_msg = Float64MultiArray()
        for node in tree:
            tree_msg.data.append(node.x)
            tree_msg.data.append(node.y)
            tree_msg.data.append(tree[node.parent].x)
            tree_msg.data.append(tree[node.parent].y)
        self.tree_pub.publish(tree_msg)

File: src/rrt.cpp
// ESE 680
// RRT assignment
// Author: Hongrui Zheng

// This file contains the class definition of tree nodes and RRT
// Before you start, please read: https://arxiv.org/pdf/1105.1186.pdf
// Make sure you have read through the header file as well

#include "rrt/rrt.h"

// CSV reader class
// Destructor
CSVReader::~CSVReader() {
    ROS_INFO("CSVReader shutting down.");
}
// Constructor
CSVReader::CSVReader(std::string file,
                     std::string delim)
    : file_name(file), delimeter(delim) {
    ROS_INFO("Starting CSVReader");
}
// CSVReader member functions
std::vector<std::vector<double>> CSVReader::get_wpts() {
    // Returns the waypoints stored in the file used to create
    // this instance of CSVReader as a std vector of std arrays.
    // You can change the size of the std arrays to keep more
    // information from the waypoint csv file. The defualt of
    // size 2 only keeps the x and y coordinate of the waypoints.
    ROS_INFO("Reading from: %s", file_name.c_str());
    std::ifstream file(file_name);
    std::vector<std::vector<double>> waypoints;
    std::string line = "";
    std::string::size_type sz;
    while(getline(file, line)) {
        std::vector<std::string> vec;
        boost::algorithm::split(vec, line, boost::is_any_of(delimeter));
        // change below if you need more information from the file
        double x = std::stod(vec[0], &sz);
        double y = std::stod(vec[1], &sz);
        std::vector<double> point {x, y};
        waypoints.push_back(point);
    }
    file.close();
    ROS_INFO("Waypoints loaded, length: %zd", waypoints.size());
    return waypoints;
}

// Destructor of the RRT class
RRT::~RRT() {
    // Do something in here, free up used memory, print message, etc.
    ROS_INFO("RRT shutting down");
}

// Constructor of the RRT class
RRT::RRT(ros::NodeHandle &nh): nh_(nh), gen((std::random_device())()) {

    // TODO: Load parameters from yaml file
    std::string path_topic, tree_topic, pf_topic, static_topic, env_topic, dynamic_topic;
    nh_.getParam("path_topic", path_topic);
    nh_.getParam("tree_topic", tree_topic);
    nh_.getParam("pf_pose_topic", pf_topic);
    nh_.getParam("static_layer_topic", static_topic);
    nh_.getParam("env_layer_topic", env_topic);
    nh_.getParam("dynamic_layer_topic", dynamic_topic);
    

    // ROS publishers
    // TODO: create publishers for the the drive topic
    
    
    path_pub_ = nh_.advertise<nav_msgs::Path>(path_topic, 10);
    tree_pub_ = nh_.advertise<std_msgs::Float64MultiArray>(tree_topic, 10);

    // ROS subscribers
    // subscribe to particle filter
    pf_sub_ = nh_.subscribe(pf_topic, 10, &RRT::pf_callback, this);
    // subscribe to occupancy grid layers
    stat_sub_ = nh_.subscribe(static_topic, 10, &RRT::static_callback, this);
    env_sub_ = nh_.subscribe(env_topic, 10, &RRT::env_callback, this);
    dynamic_sub_ = nh_.subscribe(dynamic_topic, 10, &RRT::dynamic_callback, this);

    // Wait for map metatdata message to update class member
    boost::shared_ptr<nav_msgs::MapMetaData const> env_metadata_ptr;
    env_metadata_ptr = ros::topic::waitForMessage<nav_msgs::MapMetaData>("/map_metadata");
    if (env_metadata_ptr != NULL) {
        map_resolution = env_metadata_ptr->resolution;
        map_width = env_metadata_ptr->width;
        map_height = env_metadata_ptr->height;
        origin_x = env_metadata_ptr->origin.position.x;
        origin_y = env_metadata_ptr->origin.position.y;
        ROS_INFO("Map Metadata Loaded.");
    } else {
        ROS_ERROR("Map Metadata not loaded.");
    }

    // Wait for occupancy grid messages to arrive
    boost::shared_ptr<nav_msgs::OccupancyGrid const> dynamic_ptr;
    boost::shared_ptr<nav_msgs::OccupancyGrid const> static_ptr;
    boost::shared_ptr<nav_msgs::OccupancyGrid const> env_ptr;
    dynamic_ptr = ros::topic::waitForMessage<nav_msgs::OccupancyGrid>(dynamic_topic);
    static_ptr = ros::topic::waitForMessage<nav_msgs::OccupancyGrid>(static_topic);
    env_ptr = ros::topic::waitForMessage<nav_msgs::OccupancyGrid>(env_topic);
    if (dynamic_ptr != NULL && static_ptr != NULL && env_ptr != NULL) {
        static_callback(static_ptr);
        dynamic_callback(dynamic_ptr);
        env_callback(env_ptr);
        ROS_INFO("Occupancy Grid messages arrived.");
    } else {
        ROS_ERROR("Occupancy Grid layers not loaded.");
    }
    ROS_INFO("Created new RRT Object.");
}


void RRT::pf_callback(const geometry_msgs::PoseStamped::ConstPtr &pose_msg) {
    // The pose callback when subscribed to particle filter's inferred pose
    // The RRT main loop happens here
    // Args:
    //    pose_msg (*PoseStamped): pointer to the incoming pose message
    // Returns:
    //

    // tree as std::vector
    std::vector<Node> tree;
    std::vector<Node> path;

    // TODO: fill in the RRT main loop



    // path found as Path message
    nav_msgs::Path path_msg;
    path_msg.header.frame_id = "/map";
    // TODO: fill message with your found path
    
    path_pub_.publish(path_msg);
    
    // visualization
    pub_tree(tree);
}

std::vector<double> RRT::sample() {
    // This method returns a sampled point from the free space
    // You should restrict so that it only samples a small region
    // of interest around the car's current position
    // Args:
    // Returns:
    //     sampled_point (std::vector<double>): the sampled point in free space

    std::vector<double> sampled_point;
    // TODO: fill in this method
    
    return sampled_point;
}


int RRT::nearest(std::vector<Node> &tree, std::vector<double> &sampled_point) {
    // This method returns the nearest node on the tree to the sampled point
    // Args:
    //     tree (std::vector<Node>): the current RRT tree
    //     sampled_point (std::vector<double>): the sampled point in free space
    // Returns:
    //     nearest_node (int): index of nearest node on the tree

    int nearest_node = 0;
    // TODO: fill in this method

    return nearest_node;
}

Node RRT::steer(Node &nearest_node, std::vector<double> &sampled_point) {
    // The function steer:(x,y)->z returns a point such that z is “closer” 
    // to y than x is. The point z returned by the function steer will be 
    // such that z minimizes ||z−y|| while at the same time maintaining 
    //||z−x|| <= max_expansion_dist, for a prespecified max_expansion_dist > 0

    // basically, expand the tree towards the sample point (within a max dist)

    // Args:
    //    nearest_node (Node): nearest node on the tree to the sampled point
    //    sampled_point (std::vector<double>): the sampled point in free space
    // Returns:
    //    new_node (Node): new node created from steering

    Node new_node;
    // TODO: fill in this method

    return new_node;
}

bool RRT::check_collision(Node &nearest_node, Node &new_node) {
    // This method returns a boolean indicating if the path between the 
    // nearest node and the new node created from steering is collision free
    // Args:
    //    nearest_node (Node): nearest node on the tree to the sampled point
    //    new_node (Node): new node created from steering
    // Returns:
    //    collision (bool): true if in collision, false otherwise

    bool collision = false;
    // TODO: fill in this method

    return collision;
}

bool RRT::is_goal(Node &latest_added_node, double goal_x, double goal_y) {
    // This method checks if the latest node added to the tree is close
    // enough (defined by goal_threshold) to the goal so we can terminate
    // the search and find a path
    // Args:
    //   latest_added_node (Node): latest addition to the tree
    //   goal_x (double): x coordinate of the current goal
    //   goal_y (double): y coordinate of the current goal
    // Returns:
    //   close_enough (bool): true if node close enough to the goal

    bool close_enough = false;
    // TODO: fill in this method

    return close_enough;
}

std::vector<Node> RRT::find_path(std::vector<Node> &tree, Node &latest_added_node) {
    // This method traverses the tree from the node that has been determined
    // as goal
    // Args:
    //   latest_added_node (Node): latest addition to the tree that has been
    //      determined to be close enough to the goal
    // Returns:
    //   path (std::vector<Node>): the vector that represents the order of
    //      of the nodes traversed as the found path
    
    std::vector<Node> found_path;
    // TODO: fill in this method

    return found_path;
}

// RRT* methods
double RRT::cost(std::vector<Node> &tree, Node &node) {
    // This method returns the cost associated with a node
    // Args:
    //    tree (std::vector<Node>): the current tree
    //    node (Node): the node the cost is calculated for
    // Returns:
    //    cost (double): the cost value associated with the node

    double cost = 0;
    // TODO: fill in this method

    return cost;
}

double RRT::line_cost(Node &n1, Node &n2) {
    // This method returns the cost of the straight line path between two nodes
    // Args:
    //    n1 (Node): the Node at one end of the path
    //    n2 (Node): the Node at the other end of the path
    // Returns:
    //    cost (double): the cost value associated with the path

    double cost = 0;
    // TODO: fill in this method

    return cost;
}

std::vector<int> RRT::near(std::vector<Node> &tree, Node &node) {
    // This method returns the set of Nodes in the neighborhood of a 
    // node.
    // Args:
    //   tree (std::vector<Node>): the current tree
    //   node (Node): the node to find the neighborhood for
    // Returns:
    //   neighborhood (std::vector<int>): the index of the nodes in the neighborhood

    std::vector<int> neighborhood;
    // TODO:: fill in this method

    return neighborhood;
}





// Occupancy grid callbacks, update class members that stores layer info
// The three layers of Occupancy Grid are stored as: stat, env, and dynamic
void RRT::static_callback(const nav_msgs::OccupancyGrid::ConstPtr &static_layer) {
    // ROS_INFO("updating static layer.");
    nav_msgs::OccupancyGrid new_msg;
    new_msg.header = static_layer->header;
    new_msg.info = static_layer->info;
    new_msg.data = static_layer->data;
    // update class member
    stat = new_msg;
}
void RRT::env_callback(const nav_msgs::OccupancyGrid::ConstPtr &env_layer) {
    // ROS_INFO("updating env layer.");
    nav_msgs::OccupancyGrid new_msg;
    new_msg.header = env_layer->header;
    new_msg.info = env_layer->info;
    new_msg.data = env_layer->data;
    // update class member
    env = new_msg;
}
void RRT::dynamic_callback(const nav_msgs::OccupancyGrid::ConstPtr &dynamic_layer) {
    // ROS_INFO("updating dynamic layer.");
    nav_msgs::OccupancyGrid new_msg;
    new_msg.header = dynamic_layer->header;
    new_msg.info = dynamic_layer->info;
    new_msg.data = dynamic_layer->data;
    // update class member
    dynamic = new_msg;
}
// For visualization
void RRT::pub_tree(std::vector<Node> &tree) {
    // publish the current tree as a float array topic
    // published as [n1.x, n1.y, n1.parent.x, n1.parent.y, ......]
    int tree_length = tree.size();
    std_msgs::Float64MultiArray tree_msg;
    for (int i=1; i<tree_length; i++) {
        double x = tree[i].x, y = tree[i].y;
        double px, py;
        if (tree[i].parent == -1) {
            px = 0.0;
            py = 0.0;
        } else {
            px = tree[tree[i].parent].x, py = tree[tree[i].parent].y;
        }
        tree_msg.data.push_back(x);
        tree_msg.data.push_back(y);
        tree_msg.data.push_back(px);
        tree_msg.data.push_back(py);
    }
    tree_pub_.publish(tree_msg);
}